Added access control public header file. (#1300)
Haivision/srt: docs/AccessControl.md

the listener and its name is `results.csv`.

### Rejection codes

The listener callback handler is also able to decide about rejecting the
incoming connection. In a normal situation, the rejection code is predefined
as `SRT_REJ_RESOURCE`. The handler can, however, sets its own rejection
code. There are two number spaces intended for this purpose (as the range
below `SRT_REJC_PREDEFINED` is reserved for internal codes):

* `SRT_REJC_PREDEFINED` and above: predefined errors. Errors from this range
(that is, below `SRT_REJC_USERDEFINED`) have their definitions provided in
the `access_control.h` public header file. The intention is that applications

> using these codes understood the situation described by these codes standard
> way.

Reviewer: ```suggestion
using these codes understand the situation described by these codes in a standard
way.
```

* `SRT_REJC_USERDEFINED` and above: to be freely defined by the application.

> Codes from this range can be only understood if both applications know the
> code definitions of the other, and these should be used only after making
> sure that both applications understood these codes.

Reviewer: ```suggestion
Codes from this range can be only understood if each applications knows the
code definitions of the other. These codes should be used only after making
sure that both applications understood them.
```

> The intention for the predefined codes is to be consistent with the HTTP
> standard codes, therefore there are following sub-ranges used:

Reviewer: ```suggestion
The intention for the predefined codes is to be consistent with HTTP
standard codes. Therefore the following sub-ranges are used:
```

* 0 - 99: Reserved for unique SRT-specific codes (unused by HTTP)
* 100 - 399: Info, Success and Redirection in HTTP, unused in SRT
* 400 - 599: Client and server errors in HTTP, adopted by SRT
* 600 - 999: unused in SRT

> This code can be set by using the `srt_setrejectreason` function.

Reviewer: ```suggestion
Such a code can be set by using the `srt_setrejectreason` function.
```

The SRT-specific codes are:

> ### SRT_REJX_FAlLBACK

Reviewer: ```suggestion
### SRT_REJX_FAlLBACK  <-- is this failback or fallback?
```

Author: Of course it should be FALLBACK. Damned kezboard. This has an extra key right of the left shift, so shift isn't always correctly pressed...

> This code should be set by the callback handler in the beginning in case
> when the application needs to be informed that the callback handler
> actually has interpreted the incoming connection, but it didn't set any
> more appropriate code describing the situation.

Reviewer: ```suggestion
This code should be set by the callback handler in the beginning in case 
the application needs to be informed that the callback handler
actually has interpreted the incoming connection, but hasn't set a
more appropriate code describing the situation.
```

### SRT_REJX_KEY_NOTSUP

> There was a key specified in the StreamID string that this application
> doesn't support. Note that it's not obligatory for the application to
> react this way - it may chose to ignore unknown keys completely, or
> have some keys in the ignore list, which it won't interpret, but tolerate,
> while rejecting if seen any other. It is also up to the application
> to decide to return this specific error, or more generally report
> the syntax error by `SRT_REJX_BAD_REQUEST`.

Reviewer: ```suggestion
Indicates there was a key specified in the StreamID string that this application
doesn't support. Note that it's not obligatory for the application to
react this way - it may chose to ignore unknown keys completely, or
to have some keys in the ignore list (which it won't interpret, but tolerate)
while rejecting any others. It is also up to the application
to decide to return this specific error, or more generally report
the syntax error with `SRT_REJX_BAD_REQUEST`.
```

### SRT_REJX_FILEPATH

> The resource type designates a file and the path is either wrong syntax
> or not found. In case when `t=file`, the path should be specified under
> the `r` key, and the file specified there must be able to be saved this
> way. It's up to the application to decide how to treat this path, how
> to parse it, and what this path specifically means. What should be only
> generally satisfied for the `r` key is that the application should at
> least handle the single filename and have some storage where it can
> save it in this form, provided that it doesn't have yet a file of that
> name there. How to handle all other situations - like directory path,
> special markers in the path to be interpreted by the application etc. -
> it's completely up to the application whether and how it handles it.
> If not - it should report this error.

Reviewer: ```suggestion
The resource type designates a file, and the path either has the wrong syntax
or is not found. In the case where `t=file`, the path should be specified under
the `r` key, and the file specified there must be able to be saved this
way. It's up to the application to decide how to treat this path, how
to parse it, and what this path specifically means. For the `r` key, the 
application should at least handle the single filename, and have storage 
space available to save it (provided a file of the same name does not already exist there).
The application should decide whether and how to handle all other situations (like directory path,
special markers in the path to be interpreted by the application, etc.), or to report this error.
```

### SRT_REJX_HOSTNOTFOUND

The host specified in the `h` key cannot be identified. The `h` key is
generally for a situation when you have multiple DNS names for a host,
so an application may want to extract the name from the URI and set it
to `h` key so that the application can distinguish the request also by
the target host name. The application may however limit the number of
recognized services by host name to some predefined names and not
handle the others, even if this is properly resolved by DNS. In this
case it should report this error.

The other error codes are HTTP codes adopted for SRT:

### SRT_REJX_BAD_REQUEST

General syntax error. This can be reported in any case when parsing
the StreamID contents failed, or it cannot be properly interpreted.

### SRT_REJX_UNAUTHORIZED

Authentication failed, which makes the client unauthorized to access the
resource. This error, however, confirms that the syntax is correct and
the resource has been properly identified. Note that this cannot be

> reported in case when you use a simple user-password authentication

Reviewer: ```suggestion
reported when you use a simple user-password authentication
```

method because in this case the password is verified only after the
listener callback handler accepts the connection. This error is rather
intended to be reported in case of `t=auth` when the authentication
process has generated some valid session ID, but then the session
connection has specified a resource that is not within the frames
of that authentication.

### SRT_REJX_OVERLOAD

> The server is too heavily loaded to process your request, or you
> have exceeded credits for accessing the service and the resource.
> In HTTP the description mentioned the payment for the service, but
> it was also in use by some service as more general "credit" management
> for a client. In SRT it should be used in case when your service
> is doing any kind of credit management to limit access to the service
> only to some selected clients that "have" enough credit - no matter
> if the credit is something they can recharge themselves or it can
> be also granted them depending on available service resources.

Reviewer: ```suggestion
The server is too heavily loaded to process your request, or you
have exceeded credits for accessing the service and the resource.
In HTTP the description mentions payment for a service, but
it is also used by some services for general "credit" management
for a client. In SRT it should be used when your service is doing 
any kind of credit management to limit access to selected clients 
that "have" enough credit, even if the credit is something the client 
can recharge itself, or that can be granted depending on available 
service resources.
```

### SRT_REJX_FORBIDDEN

> Access denied to the resource by any kind of reason. This error is
> independet on an authorization or authentication error (as reported
> by `SRT_REJX_UNAUTHORIZED`), and whichever is more appropriate here is up to
> the decision of the application. This error is usually intended for
> a situation when the resource should be only accessed after a successful
> authorization over a separate auth-only connection, while the query
> in StreamID has correctly specified the resource identity and mode,
> but the session ID (in `s` key) is either not specified or it does
> specify a valid session, however the authorization region for this
> session does not embrace the specified resource.

Reviewer: ```suggestion
Access denied to the resource for any reason. This error is
independent of an authorization or authentication error (as reported
by `SRT_REJX_UNAUTHORIZED`). The application can decide which 
is more appropriate. This error is usually intended for
a resource that should only be accessed after a successful
authorization over a separate auth-only connection, where the query
in StreamID has correctly specified the resource identity and mode,
but the session ID (in the `s` key) is either (a) not specified, or (b) does
specify a valid session, but the authorization region for this
session does not embrace the specified resource.
```

### SRT_REJX_NOTFOUND

> The resource specified in the `r` key (with combination of `h` key)
> is not found at this time. This error should be only reported if the
> information about resource accessibility is allowed to be publicly
> visible, otherwise the application might rather report authorization
> errors.

Reviewer: ```suggestion
The resource specified in the `r` key (in combination with the `h` key)
is not found at this time. This error should be only reported if the
information about resource accessibility is allowed to be publicly
visible. Otherwise the application might report authorization
errors.
```

### SRT_REJX_BAD_MODE

> The mode specified in the `m` key in StreamID is not supported for this request.
> This is in case of having read-only or write-only resources, as well as the
> interactive (bidirectional) access is not predicted as a valid access to this
> resource.

Reviewer: ```suggestion
The mode specified in the `m` key in StreamID is not supported for this request.
This may apply to read-only or write-only resources, as well for when interactive 
(bidirectional) access is not valid for a resource.
```

### SRT_REJX_UNACCEPTABLE

> The requested parameters specified in SocketID cannot be satisfied for the
> requested resource. Also when m=publish and the data format is not acceptable.
> This is a general error reporting the unsupported data format for the data
> in case when it appears to be wrong when sending, or a restriction on the
> data format (as specified in the details of the resource specification) cannot
> be provided when receiving.

Reviewer: ```suggestion
Applies when the parameters specified in SocketID cannot be satisfied for the
requested resource, or when `m=publish` but the data format is not acceptable.
This is a general error reporting an unsupported format for data that appears to 
be wrong when sending, or a restriction on the data format (as specified in the 
details of the resource specification) such that it cannot be provided 
when receiving.
```

### SRT_REJX_CONFLICT

The resource being accessed is already locked for modification. This error

> should only be reported for `m=publish` in case when the resource being

Reviewer: ```suggestion
should only be reported for `m=publish` when the resource is being
```

accessed is read-only. Note that it doesn't mean that the resource is not
modifiable under current authorization restrictions, but that it is currently
disabled completely for modifications.

### SRT_REJX_NOTSUP_MEDIA

The media type is not supported by the application. The media type is

> specified in the `t` key, although the currently standard types are
> `stream`, `file` and `auth`, an application may extend this list, and
> also it's not obliged to support all of the standard ones.

Reviewer: ```suggestion
specified in the `t` key. The currently standard types are
`stream`, `file` and `auth`. An application may extend this list, and
is not obliged to support all of the standard types.
```

### SRT_REJX_LOCKED

> The resource being accessed is locked for any access. This is similar to
> `SRT_REJX_CONFLICT`, but in this case it's locked for reading and writing.
> This is for a case when the resource should be shown as existing and in
> general available for the client, just temporarily blocked for access at
> the moment.

Reviewer: ```suggestion
The resource being accessed is locked against any access. This is similar to
`SRT_REJX_CONFLICT`, but in this case the resource is locked for reading 
and writing. This is for when the resource should be shown as existing and 
available to the client, but access is temporarily blocked.
```

### SRT_REJX_FAILED_DEPEND 

The dependent entity for the request is not present. In this case the
dependent entity is the session, which should be specified in the `s`
key. This means that the specified session ID is nonexistent or it
has already expired.

### SRT_REJX_ISE

> Internal server error. This is for a general case when the request has
> been correctly verified and no problems related to this have been found,
> the processing of the request has started and the activity was about to
> be correctly performed, just there was some unexpected error during this
> process.

Reviewer: ```suggestion
Internal server error. This is for a general case when a request has
been correctly verified, with no related problems found, but an 
unexpected error occurs after the processing of the request has started.
```

### SRT_REJX_UNIMPLEMENTED

> The request was correctly recognized, but the current version doesn't

Reviewer: "current version" of what? SRT? The application? Both? Later you refer to a "server version". How does this relate to SRT and/or the application?

Author: It's all about the server version. The rejection codes concern exclusively a situation when you have a service hanging as "SRT Listener" and a client tries to connect, with having a query specified in the StreamID field. Although yes, it should be clarified that it's about the version of the server application (can be both SRT or any other part of the service software).

support it. This should be reported for a case, when some features to
be specified in the StreamID request are supposed to be supported in a
predictable feature, but the current version of the server does not
support it, or the support for this feature in this version has been
temporarily blocked. This shouldn't be reported in case of deprecated

> features or features that were decided as no longer supported.

Reviewer: What is the difference between "deprecated features" and
"features that were decided as no longer supported"?

Author: "deprecated" is such that the request for it is tolerated as a kinda backward compatibility (specifying them is not an error), but they may generate a warning, or it's announced in the documentation that this feature is about to be removed in future versions. "features that were decided as no longer supported" are those that are simply not supported (and trying to use them results in an error), although they were supported in the earlier version, or more precisely, in the previous major version they were deprecated.

Reviewer: Suggested edit: "This shouldn't be reported for existing features that are being deprecated, or older features that are no longer supported.

### SRT_REJX_GW

> The server acts as a gateway and the target endpoint rejected the
> connection. It is unspecified by what exactly reason the connection
> was rejected. The gateway shall not forward the original rejection
> code from the target endpoint because this would suggest the error
> on the gateway itself - rather then this error, with possibly some
> other way to report the original target error, if there is a possibility.

Reviewer: ```suggestion
The server acts as a gateway and the target endpoint rejected the
connection. The reason the connection was rejected is unspecified. 
The gateway cannot forward the original rejection code from the 
target endpoint because this would suggest the error was on the 
gateway itself. Use this error with some other mechanism to report 
the original target error, if possible.
```

### SRT_REJX_DOWN

> The service is down for maintenance. This can only be reported in
> case when the servis has been temporarily replaced by a stub that is only
> reporting this error, while the real service is down for maintenance.

Reviewer: ```suggestion
The service is down for maintenance. This can only be reported 
when the service has been temporarily replaced by a stub that is only
reporting this error, while the real service is down for maintenance.
```

### SRT_REJX_VERSION

> Application version not supported. This can refer to both not supported
> application feature - also due to too low SRT version - or about referring
> to a feature that is no longer supported as a cut off backward comptibility.

Reviewer: ```suggestion
Application version not supported. This can refer to an application feature 
that is unsupported (possibly from an older SRT version), or to a feature 
that is no longer supported because of backward compatibility requirements.
```

### SRT_REJX_NOROOM

> The data stream cannot be archived due to lacking storage space. This is
> reported in case when the request type was to send a file or a live stream
> to be archived. Note that a file transmission is usually pre-declaring the
> file lentgh, so this error can be reported early, but it can be as well
> reported when the stream is of undefined length and there's no more storage
> space made available for this stream.

Reviewer: ```suggestion
The data stream cannot be archived due to a lack of storage space. This is
reported a request to send a file or a live stream to be archived is unsuccessful. 
Note that the length of a file transmission is usually pre-declared, so this error 
can be reported early. It can also be reported when the stream is of undefined 
length, and there is no more storage space available.
```

Author: I think there's lacking one word here.

`This is reported` **when** `a request to send a file or a live stream to be archived is unsuccessful` ?

Reviewer: Yes, I missed that.
